### PR TITLE
Add ability to run block tests against specific Gutenberg versions

### PR DESCRIPTION
**Pull request by glendaviesnz**
#### Changes proposed in this Pull Request:
* Adds a script to allow running of block unit and fixture tests against specific versions of gutenberg

#### Does this pull request change what data or activity we track or use?
No

#### Testing instructions:
* Checkout PR  
* Run `pnpm fixtures:test:gbv trunk 3774b1489f17de1da53a96a2066ce9a8f996293c`
* This will check out the commit that caused the recent 10.9.0 regressions with the subscriptions block, so once this task has completed you should see 2 block fixture failures
* Run `pnpm fixtures:test:gbv test-validation` to just run the tests again
* Run `pnpm fixtures:test:gbv reset` and check that `./packages` and `./tests/temp-gutenberg-checkout` dirs have been removed and the `./package.json` changes have been reverted and `pnpm test-extensions` completes successfully
* Run `pnpm fixtures:test:gbv trunk` and make sure all the tests complete successfully

#### Todo:

- [ ] Get block unit tests running

### CODE REVIEW

**Comment by brbrr** [comment]
@glendaviesnz Is there any info where I can learn more about the motivation behind this PR?

**Comment by glendaviesnz** [comment]
> @glendaviesnz Is there any info where I can learn more about the motivation behind this PR?

p1HpG7-cuI-p2

File: projects/plugins/jetpack/tests/test-gutenberg-version.mjs [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+		[
+			'add',
+			'showdown',
+			'simple-html-tokenizer',
+			'hpq',
+			'react-autosize-textarea',
+			'traverse',
+			'css-mediaquery',
+			'@emotion/styled',
+			'@emotion/react',
```
Comment (anomiex) [inline projects/plugins/jetpack/tests/test-gutenberg-version.mjs:159]: What's going on here? Are all of these going to have to be added someday when we get around to making #19328 happen?

**Review by anomiex — COMMENTED** [review]
I think the idea of being able to run against a different version of Gutenberg is a good idea, but I don't like the amount of hackery going on in the script you're adding here.

I have to wonder how this even works given the failures in #19328 where renovate is trying to update the same pacakges.

Unfortunately I don't know enough about how all this stuff works to be able to suggest a better way to do things.

File: projects/plugins/jetpack/package.json [hunk 2]
```diff
@@ -38,6 +38,7 @@
 		"fixtures:regenerate": "REGENERATE_FIXTURES=y GENERATE_MISSING_FIXTURES=y pnpx jest",
 		"fixtures:generate": "GENERATE_MISSING_FIXTURES=y pnpx jest",
 		"fixtures:test": "jest validate.js",
+		"fixtures:test:gbv": "node tests/test-gutenberg-version.mjs",
```
[the remark this answers is not in the document]
Comment (glendaviesnz) [reply projects/plugins/jetpack/package.json:41]: Good idea, will do that if we decide to go ahead with this PR based on outcome of discussions at p1HpG7-cuI-p2

File: projects/plugins/jetpack/tests/test-gutenberg-version.mjs (hunk 1 above)
Comment (glendaviesnz) [reply projects/plugins/jetpack/tests/test-gutenberg-version.mjs:159]: This is just temporarily adding some additional dependencies that the latest Gutenberg needs to successfully build - I assume that these could be removed when #19328 happens, although there is always the potential that the cutting edge Gutenberg is going to have some dependencies that the Jetpack repo doesn't.

**Comment by glendaviesnz** [comment]
> I think the idea of being able to run against a different version of Gutenberg is a good idea, but I don't like the amount of hackery going on in the script you're adding here.

Not the ideal setup in my view either, but I couldn't find any other way of making this happen without a temporary adjustment of the current dependencies - and the only way to automate that is a temporary hack of the package.json files as far as I can see. 

If the issues with #19328 are resolved would the plan be to try and keep the Jetpack repo more in-sync with the latest @wordpress package releases? If that was the case maybe this approach isn't necessary? 

> I have to wonder how this even works given the failures in #19328 where renovate is trying to update the same pacakges.

It probably works because we aren't trying to build the whole project with the new dependencies, it is only the Jetpack block code that is being compiled against them on the fly via the jasmine test runner

**Comment by anomiex** [comment]
> If the issues with #19328 are resolved would the plan be to try and keep the Jetpack repo more in-sync with the latest @WordPress package releases? If that was the case maybe this approach isn't necessary?

It'd be good to keep better on top of the [renovate backlog](https://github.com/Automattic/jetpack/issues/19284), yes. Part of the problem is that we don't have anyone really familiar with all this React code on Crew, so if something breaks in an update it's hard for us to fix on our own.

> It probably works because we aren't trying to build the whole project with the new dependencies, it is only the Jetpack block code that is being compiled against them on the fly via the jasmine test runner

But even `test-extensions` [is failing there](https://github.com/Automattic/jetpack/pull/19328/checks?check_run_id=2957692945#step:16:3518).

**Comment by glendaviesnz** [comment]
> But even `test-extensions` [is failing there](https://github.com/Automattic/jetpack/pull/19328/checks?check_run_id=2957692945#step:16:3518).

Yeh, because of changes in the Gutenberg packages a few of the tests were failing, I had to update quite a few of the tests to get them to run against the different dependency trees. 

In theory, if I get this PR finalised and merged, and #19328 is rebased, the test-extensions issues should be resolved there.

**Comment by anomiex** [comment]
> If the issues with #19328 are resolved would the plan be to try and keep the Jetpack repo more in-sync with the latest @WordPress package releases? If that was the case maybe this approach isn't necessary?

> Yeh, because of changes in the Gutenberg packages a few of the tests were failing, I had to update quite a few of the tests to get them to run against the different dependency trees.
>
> In theory, if I get this PR finalised and merged, and #19328 is rebased, the test-extensions issues should be resolved there.

Maybe it'd make more sense to fix #19328 so it can be merged, then see if the controversial parts of this approach are still necessary?

That might also simplify the necessary changes, as when doing in #19328 you wouldn't need to continue to support the old versions of the packages too.

**Comment by glendaviesnz** [comment]
> Maybe it'd make more sense to fix #19328 so it can be merged, then see if the controversial parts of this approach are still necessary?

Yeh, good idea, I will take another look at it from that angle.

**Comment by glendaviesnz** [comment]
Closed in favour of https://github.com/Automattic/jetpack/pull/20327 which removes the test updates which are moved to https://github.com/Automattic/jetpack/pull/20326

**Pull request closed without merging**